### ISSUE TITLE
[Feat] Ability to change tiles of raster <Source>

### ISSUE THREAD

**Issue by fungiboletus** [visgl/react-map-gl#1876]
### Target Use Case

I have a `<Source>` of type `raster` in which I would like to change the tiles dynamically.

My use case is that my source is a WMS satellite pictures API of an area in which I can select the date (when it's not too cloudy), or the visualisation type (natural colours, temperature, reflection…).

Updating tiles of raster source is unsupported as it throws an exception `Unable to update <Source> prop: tiles` in `source.ts`.

I tried to create many sources at first and switch them using the `source` attribute of `<Layer>` but it doesn't seem to be updated.

### Proposal

Source of type `raster` could support updating their tiles.

**Comment by fungiboletus** [comment]
The `raster` source look a bit undocumented in MapboxGLJS, the documentation has an example but not much more.

**Comment by fungiboletus** [comment]
My workaround is to add a `key` property to my source and layer that change when the `tiles` property change, so the source and layer gets recreated instead of being updated.

**Comment by shermify** [comment]
I don't think you can reliably change the tiles url of a source in mapbox.  Even if you do, I don't think you can get the behavior you are probably looking for where the old tiles are replaced incrementally with the new tiles.  Best bet is to  recreate the source like you are doing, that way the map will automatically clear the old tiles and repaint.  I'm guessing this would be a breaking change because people might be relying on the current behavior.

**Comment by Pessimistress** [comment]
react-map-gl does not implement any source on its own. This can only be supported if Mapbox GL JS exposes a public method to update the source, similar to [VectorTileSource.setTiles](https://docs.mapbox.com/mapbox-gl-js/api/sources/#vectortilesource).

I suggest you open a feature request in the Mapbox GL JS repo if the current solution does not meet your needs.

**Comment by fungiboletus** [comment]
I created a feature request in the mapbox gl js repository.

**Comment by dhairyasenjaliya** [comment]
I want to added raster image with their own 4 coordinates

**Comment by dhairyasenjaliya** [comment]
something like this 
https://github.com/nitaliano/react-native-mapbox-gl/blob/master/docs/RasterSource.md

**Comment by shermify** [comment]
Yes https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/#image

**Comment by elietech** [comment]
> My workaround is to add a `key` property to my source and layer that change when the `tiles` property change, so the source and layer gets recreated instead of being updated.

@fungiboletus I can't seem to make this work-around work...can you point me to an example?  I am getting the following error: ```Error: Source "radarsource" cannot be removed while layer "jsx-layer-1" is using it.```

**Comment by roman-hatz-srf** [comment]
@fungiboletus we have found a way to change tiles dynamically but the transitions are too slow.

**Comment by jo-chemla** [comment]
PR merged to main in mapbox-gl-js on Feb 2, 2023 to add `setTiles` and `setURL` on a raster layer source - see [here](https://github.com/mapbox/mapbox-gl-js/pull/12352)
So one can now do, for example in a `useEffect` method which watches/subscribes to the TMS source URL property: 
```js
  useEffect(() => {
    if (mapRef.current) {
      const rasterTileSource = mapRef.current.getSource("source-id");
      rasterTileSource.setTiles([sourceTmsUrl]);
    }
  }, [sourceTmsUrl]);```

**Comment by jo-chemla** [comment]
What would be really interesting now is a way to avoid tile popping on layer tiles change. 
When one does use `setTiles` with a single TMS url, the previous rasterLayer tiles of the source are dumped - resulting in the map background style being displayed for a short amount of time until the new tiles are loaded and start populating the map.
This makes for a harsh transition especially for periodic tilesets. But this is probably more a request for [mapbox-gl-js](https://github.com/mapbox/mapbox-gl-js)

https://github.com/visgl/react-map-gl/assets/16822841/f38999eb-7084-4afb-aaa3-5e0ff388a81d

**Comment by jo-chemla** [comment]
My understanding is it's possible this was implemented as part of the react-map-gl v8 release as seen in this [blame](https://github.com/visgl/react-map-gl/blame/948fd2e4f4156756d2fcee8c2a900cfe55b43608/modules/react-mapbox/src/components/source.ts#L71), but the behavior is still the same as back then, even hwen no key are set on the react map source or layer, and only the tiles url is modified from one date to another. Still all tiles are unloaded at once before each tile of the new tileset gets loaded. Might be related to these non intuitive map events fired discussed [here](https://github.com/visgl/react-map-gl/pull/2524/files#r2033835529)?
